### DISCUSSION TITLE
How to Add Custom Predicate Support to Apache Jena Fuseki?

### DISCUSSION THREAD

**Discussion by Myasir92** [apache/jena#2377] [General]
I'm using the Apache Jena Fuseki executable locally for data management, but I've noticed it lacks support for a valuable predicate from Sesame RDF4J. My goal is to incorporate this support by adding custom rules. While it's not as straightforward as using RDF4J directly, I am familiar with the predicate's rules and believe adding these custom rules to Fuseki is possible. Could you guide me on how to effectively add custom rules in Fuseki?

**Comment by ajs6f** [comment]
Hi @Myasir92, I'm not quite sure what you mean by "support for a ... predicate". Of which predicate are you thinking? Do you mean simply the use of the predicate? You should be able to use any well-formed predicate right away. Or is there a special semantic for this predicate in RDF4J that you would like to see enforced or enacted in your Fuseki instance? Please give us a little more context.

**Reply by ajs6f** [reply]
Okay, I'm now reading your question more carefully, and I see that you refer to rules. This is something that Jena [can handle very well](https://jena.apache.org/documentation/inference/#rules), and if you give us some specifics we can advise you more fully.

**Comment by Myasir92** [comment]
Hi @ajs6f I appreciate your quick response. I'd like to implement a feature similar to the 'directType' predicate of sesame RDF4J, which is supported in other triplestores like GraphDB. This feature is quite useful, but currently not available in Fuseki. My plan is to write a custom rule to enable this support within Fuseki.

**Reply by ajs6f** [reply]
Is [this](https://rdf4j.org/documentation/reference/configuration/#direct-type-inferencer) the inferencing regime you want to implement? IIUC you want to be able to query the explicitly declared `rdf:type` of entities without querying the inferred types?

**Reply by Myasir92** [reply]
@ajs6f Yeah

**Comment by afs** [comment]
https://jena.apache.org/documentation/inference/#directRelations

and I found this in the code which shows the way to name direct prediciates
https://github.com/apache/jena/blob/4fb3ce2c34531dcbab8afba7b5595038b4d93d39/jena-core/src/main/java/org/apache/jena/reasoner/ReasonerRegistry.java#L177

[Myasir92](https://github.com/Myasir92) -

This is one time when you may find the `users@jena.apache.org` useful because some people with expertise in the Jena rule system only have time to read and response there.

**Reply by Myasir92** [reply]
@afs Could you please provide an example of a SPARQL query that utilizes that predicate, or are there additional steps that need to be taken?

**Comment by ajs6f** [comment]
Thank you @afs! 

This method on `OntResource` may also be useful (see the parameter):

https://github.com/apache/jena/blob/4fb3ce2c34531dcbab8afba7b5595038b4d93d39/jena-core/src/main/java/org/apache/jena/ontology/OntResource.java#L533
